### PR TITLE
feat: add Cloud Monitoring MQL backend

### PR DESCRIPTION
**Pull request by lvaylet**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -63,6 +63,9 @@ develop:
 install: clean
 	$(PIP) install -e ."[api, datadog, prometheus, elasticsearch, pubsub, cloud_monitoring, bigquery, dev]"
 
+uninstall: clean
```
Comment (lvaylet) [inline Makefile:66]: I added this target so I could test different versions of the `google-cloud`monitoring` package and make sure I start from a clean slate.

File: samples/cloud_monitoring_mql/slo_gae_app_availability_ratio.yaml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+            || metric.response_code == 203
+            || metric.response_code == 204
+            || metric.response_code == 205
+            || metric.response_code == 206
+            || metric.response_code == 207
+            || metric.response_code == 208
+            || metric.response_code == 226
+            || metric.response_code == 304
+        ; ident }
+      | ratio
```
Comment (lvaylet) [inline samples/cloud_monitoring_mql/slo_gae_app_availability_ratio.yaml:34]: This sample uses the built-in `ratio` function of MQL to compute the SLI directly. A `query_sli` method is responsible for computing this, like in the Prometheus backend.

File: setup.py [hunk 3]
```diff
@@ -89,4 +89,4 @@
       entry_points={
           'console_scripts': ['slo-generator=slo_generator.cli:main'],
       },
-      python_requires='>=3.4')
+      python_requires='>=3.6')
```
Comment (lvaylet) [inline setup.py:92]: As required by `google-cloud-monitoring` v2.

File: slo_generator/backends/cloud_monitoring.py [hunk 4]
```diff
@@ -87,10 +87,10 @@ def good_bad_ratio(self, timestamp, window, slo_config):
         LOGGER.debug(f'Good events: {good_event_count} | '
                      f'Bad events: {bad_event_count}')
 
-        return (good_event_count, bad_event_count)
+        return good_event_count, bad_event_count
```
Comment (lvaylet) [inline slo_generator/backends/cloud_monitoring.py:90]: My linter kept complaning about redundant parentheses.

File: slo_generator/backends/cloud_monitoring.py [hunk 5]
```diff
@@ -180,15 +180,20 @@ def query(self,
...
         measurement_window = CM.get_window(timestamp, window)
         aggregation = CM.get_aggregation(window,
                                          aligner=aligner,
                                          reducer=reducer,
                                          group_by=group_by)
-        timeseries = self.client.list_time_series(
-            self.parent, filter, measurement_window,
-            monitoring_v3.enums.ListTimeSeriesRequest.TimeSeriesView.FULL,
-            aggregation)
+        request = monitoring_v3.ListTimeSeriesRequest()
```
Comment (lvaylet) [inline slo_generator/backends/cloud_monitoring.py:190]: This is a good example of how I had to rewrite the existing code for `google-cloud-monitoring` v2.

File: slo_generator/backends/cloud_monitoring.py [hunk 6]
```diff
@@ -220,20 +225,28 @@ def get_window(timestamp, window):
         Returns:
             :obj:`monitoring_v3.types.TimeInterval`: Measurement window object.
         """
-        measurement_window = monitoring_v3.types.TimeInterval()
-        measurement_window.end_time.seconds = int(timestamp)
-        measurement_window.end_time.nanos = int(
-            (timestamp - measurement_window.end_time.seconds) * 10**9)
-        measurement_window.start_time.seconds = int(timestamp - window)
-        measurement_window.start_time.nanos = measurement_window.end_time.nanos
+        end_time_seconds = int(timestamp)
```
Comment (lvaylet) [inline slo_generator/backends/cloud_monitoring.py:228]: Same here. I took these snippets from the official repo:
- https://github.com/googleapis/python-monitoring/blob/main/samples/snippets/v3/cloud-client/quickstart.py
- https://github.com/googleapis/python-monitoring/blob/main/samples/snippets/v3/cloud-client/snippets.py

File: slo_generator/backends/cloud_monitoring_mql.py [hunk 7]
```diff
@@ -0,0 +1,230 @@
+# Copyright 2019 Google Inc.
```
Comment (lvaylet) [inline slo_generator/backends/cloud_monitoring_mql.py:1]: At some point last year, we discussed augmenting the existing Cloud Monitoring backend in order to support MQL. However, I found that MQL uses a different client and different methods. I did not want to make the code unreadable by wrapping all the business logic in `if flag == MQF:`, `elif flag == MQL:`, `elif flag == PromQL`, `else ...` blocks in every method of the class. I ended up defining a new backend entirely. What do you think? The biggest drawback is probably the amount of duplicate code shared by both backends. We could centralize it somewhere else, for example in a separate library.

File: slo_generator/backends/cloud_monitoring_mql.py [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+
+        Args:
+            timestamp (int): UNIX timestamp.
+            window (int): Window (in seconds).
+            slo_config (dict): SLO configuration.
+
+        Returns:
+            float: SLI value.
+        """
+        # FIXME This method returns 2.0 for the 12-hour, 7-day and 28-day burn
```
Comment (lvaylet) [inline slo_generator/backends/cloud_monitoring_mql.py:179]: Not sure what is going wrong here. I will investigate further over the next few days. This is still a WIP.

File: slo_generator/backends/cloud_monitoring_mql.py [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+            window (int): Window size (in seconds).
+
+        Returns:
+            list: List of timeseries objects.
+        """
+        # Enrich query to aggregate and reduce the time series over the
+        # desired window.
+        # FIXME Do this properly in a private helper function. Draw inspiration
+        #  from `_fmt_query` in the Prometheus backend.
+        query += f"| group_by [] | every {window}s | within {window}s"
```
Comment (lvaylet) [inline slo_generator/backends/cloud_monitoring_mql.py:203]: My biggest concern is this line. Is it the right way to replace the aggregation and reduction performed in the Cloud Monitoring backend? I have not found anything similar in MQL (yet).

File: slo_generator/backends/cloud_monitoring_mql.py (hunk 8 above)
Comment (lvaylet) [reply slo_generator/backends/cloud_monitoring_mql.py]: I found the root cause. I need to aggregate (sum) the time series **before** computing the `ratio`.

File: slo_generator/backends/cloud_monitoring.py [hunk 10]
```diff
@@ -41,7 +41,7 @@ def __init__(self, project_id, client=None):
         self.client = client
         if client is None:
             self.client = monitoring_v3.MetricServiceClient()
-        self.parent = self.client.project_path(project_id)
+        self.parent = self.client.common_project_path(project_id)
```
Comment (lvaylet) [inline slo_generator/backends/cloud_monitoring.py:44]: One of the changes introduced by `google-cloud-monitoring` v2.

File: slo_generator/constants.py [hunk 11]
```diff
@@ -18,25 +18,25 @@
 import os
 
 # Compute
-NO_DATA = -1
-MIN_VALID_EVENTS = int(os.environ.get("MIN_VALID_EVENTS", "1"))
+NO_DATA: int = -1
```
Comment (lvaylet) [inline slo_generator/constants.py:21]: I only annotated `NO_DATA` at first, then found the file was small enough that we can annotate everything in it in this PR.

File: .gitignore [hunk 12]
```diff
@@ -43,6 +43,6 @@ htmlcov/
 *.tfstate.*.backup
 .vscode
 .env
-venv/
-.venv/
+venv*/
```
Comment (lvaylet) [inline .gitignore:46]: I wanted to support multiple local venvs with different versions of Python, for example `venv3.9.10` and `venv3.10.7`.

File: slo_generator/backends/cloud_monitoring.py [hunk 13]
```diff
@@ -220,20 +225,28 @@ def get_window(timestamp, window):
...
+        })
         LOGGER.debug(pprint.pformat(measurement_window))
         return measurement_window
 
     @staticmethod
     def get_aggregation(window,
                         aligner='ALIGN_SUM',
                         reducer='REDUCE_SUM',
-                        group_by=[]):
+                        group_by=None):
```
Comment (lvaylet) [inline slo_generator/backends/cloud_monitoring.py:249]: Default mutable argument again.

File: slo_generator/backends/cloud_monitoring_mql.py [hunk 14]
```diff
@@ -0,0 +1,262 @@
...
+        project_id (str): Cloud Monitoring host project id.
+        client (google.cloud.monitoring_v3.QueryServiceClient, optional):
+            Existing Cloud Monitoring Query client. Initialize a new client
+            if omitted.
+    """
+
+    def __init__(self, project_id: str, client: QueryServiceClient = None):
+        self.client = client
+        if client is None:
+            self.client = QueryServiceClient()
```
Comment (lvaylet) [inline slo_generator/backends/cloud_monitoring_mql.py:50]: The dedicated MQL client.

File: slo_generator/backends/cloud_monitoring_mql.py [hunk 15]
```diff
@@ -0,0 +1,262 @@
...
+        Returns:
+            str: Formatted query.
+        """
+        formatted_query: str = query.strip()
+        if 'group_by' in formatted_query:
+            formatted_query = re.sub(r'\|\s+group_by\s+\[.*\]\s*',
+                                     '| group_by [] ', formatted_query)
+        else:
+            formatted_query += '| group_by [] '
+        for mql_time_interval_keyword in ['within', 'every']:
```
Comment (lvaylet) [inline slo_generator/backends/cloud_monitoring_mql.py:251]: This code can be hard to read and maintain. Perhaps I should add some practical examples of inputs and outputs. For example the queries I used in the unit tests.

File: tests/unit/backends/test_cloud_monitoring_mql.py [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+
+# flake8: noqa
+
+import unittest
+
+from slo_generator.backends.cloud_monitoring_mql import \
+    CloudMonitoringMqlBackend
+
+
+class TestCloudMonitoringMqlBackend(unittest.TestCase):
```
Comment (lvaylet) [inline tests/unit/backends/test_cloud_monitoring_mql.py:23]: I was not sure how and where to put these unit tests in the existing suite. I wrote a dedicated ets file for the new backend.

File: tests/unit/test_stubs.py [hunk 17]
```diff
@@ -164,11 +164,11 @@ def mock_sd(nresp=1):
     Returns:
         ChannelStub: Mocked gRPC channel stub.
     """
-    timeserie = load_fixture('time_series_proto.json')
-    response = {"next_page_token": "", "time_series": [timeserie]}
+    timeseries = load_fixture('time_series_proto.json')
```
Comment (lvaylet) [inline tests/unit/test_stubs.py:167]: Just a typo.

**Review by lvaylet — COMMENTED** [review]
Mote comments.

**Review by ocervell — APPROVED** [review]
LGTM, great work !

**Pull request merged**